libbpf-tools/fsslower.c

```diff
@@ -0,0 +1,458 @@
+{
+	long err = 0;
+	struct fs_config *cfg = &fs_configs[fs_type];
+
+	/* READ */
+	obj->links.file_read_entry = bpf_program__attach_kprobe(obj->progs.file_read_entry, false, cfg->op_funcs[READ]);
+	err = libbpf_get_error(obj->links.file_read_entry);
+	if (err)
+		goto errout;
+	obj->links.file_read_exit = bpf_program__attach_kprobe(obj->progs.file_read_exit, true, cfg->op_funcs[READ]);
+	err = libbpf_get_error(obj->links.file_read_exit);
+	if (err)
+		goto errout;
+	/* WRITE */
+	obj->links.file_write_entry = bpf_program__attach_kprobe(obj->progs.file_write_entry, false, cfg->op_funcs[WRITE]);
+	err = libbpf_get_error(obj->links.file_write_entry);
+	if (err)
+		goto errout;
+	obj->links.file_write_exit = bpf_program__attach_kprobe(obj->progs.file_write_exit, true, cfg->op_funcs[WRITE]);
+	err = libbpf_get_error(obj->links.file_write_exit);
+	if (err)
+		goto errout;
+	/* OPEN */
+	obj->links.file_open_entry = bpf_program__attach_kprobe(obj->progs.file_open_entry, false, cfg->op_funcs[OPEN]);
+	err = libbpf_get_error(obj->links.file_open_entry);
+	if (err)
+		goto errout;
+	obj->links.file_open_exit = bpf_program__attach_kprobe(obj->progs.file_open_exit, true, cfg->op_funcs[OPEN]);
+	err = libbpf_get_error(obj->links.file_open_exit);
+	if (err)
+		goto errout;
+	/* FSYNC */
+	obj->links.file_sync_entry = bpf_program__attach_kprobe(obj->progs.file_sync_entry, false, cfg->op_funcs[FSYNC]);
+	err = libbpf_get_error(obj->links.file_sync_entry);
+	if (err)
+		goto errout;
+	obj->links.file_sync_exit = bpf_program__attach_kprobe(obj->progs.file_sync_exit, true, cfg->op_funcs[FSYNC]);
+	err = libbpf_get_error(obj->links.file_sync_exit);
+	if (err)
+		goto errout;
+	return 0;
+
+errout:
+	warn("failed to attach kprobe: %ld\n", err);
+	return err;
+}
+
+static void print_headers()
+{
+	const char *fs = fs_configs[fs_type].fs;
+
+	if (csv) {
+		printf("ENDTIME_ns,TASK,PID,TYPE,BYTES,OFFSET_b,LATENCY_us,FILE\n");
+		return;
+	}
+
+	if (min_lat_ms)
+		printf("Tracing %s operations slower than %llu ms", fs, min_lat_ms);
+	else
+		printf("Tracing %s operations", fs);
+
+	if (duration)
+		printf(" for %ld secs.\n", duration);
+	else
+		printf("... Hit Ctrl-C to end.\n");
+
+	printf("%-8s %-16s %-7s %1s %-7s %-8s %7s %s\n",
+	       "TIME", "COMM", "PID", "T", "BYTES", "OFF_KB", "LAT(ms)", "FILENAME");
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	const struct event *e = data;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+
+	if (csv) {
+		printf("%lld,%s,%d,%c,", e->end_ns, e->task, e->pid, file_op[e->op]);
+		e->size == LLONG_MAX ? printf("LL_MAX,") : printf("%ld,", e->size);
+		printf("%lld,%lld,%s\n", e->offset, e->delta_us, e->file);
+		return;
+	}
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+
+	printf("%-8s %-16s %-7d %c ", ts, e->task, e->pid, file_op[e->op]);
+	e->size == LLONG_MAX ? printf("%-7s ", "LL_MAX") : printf("%-7ld ", e->size);
```

Comment:
this is an abuse of ternary operator, please use proper if/else here

Reply:
ping?

Reply:
Sorry, I missed this. Use if/else instead.

Reply:
I wonder any rationale about this ?

Reduce readability ?

Reply:
ternary  operator is supposed to be used for returning expressions, it's not really a replacement for if/else statement; it's just confusing to read and easy to miss

Reply:
OK. Thanks.